abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,1471 @@
+
+Returns `InvalidMemoryAccess` when `queue_name_data`, `queue_name_size` and/or
+`return_queue_id` point to invalid memory address.
+
+#### `proxy_dequeue_shared_queue_item`
+
+* params:
+  - `i32 (uint32_t) queue_id`
+  - `i32 (iovec_array_t*) return_payload_data`
+  - `i32 (size_t*) return_payload_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Removes item from the tail of the queue (`queue_id`). Returned value is written
+into memory slice (`payload_data`, `payload_size`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the queue (`queue_id`) doesn't exist.
+
+Returns `InvalidMemoryAccess` when `return_payload_data` and/or
+`return_payload_size` point to invalid memory address.
+
+#### `proxy_enqueue_shared_queue_item`
+
+* params:
+  - `i32 (uint32_t) queue_id`
+  - `i32 (iovec_array_t) payload_data`
+  - `i32 (size_t) payload_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Adds item (`payload_data`, `payload_size`) at the front of the queue
+(`queue_id`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the queue (`queue_id`) doesn't exist.
+
+Returns `InvalidMemoryAccess` when `payload_data` and/or `payload_size` point to
+invalid memory address.
+
+#### `proxy_delete_shared_queue`
+
+* params:
+  - `i32 (uint32_t) queue_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Deletes previously created shared queue (`queue_id`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the queue (`queue_id`) doesn't exist.
+
+### Callbacks optionally exposed by the Wasm module
+
+#### `proxy_on_queue_ready`
+
+* params:
+  - `i32 (uint32_t) queue_id`
+* returns:
+  - none
+
+Called when new item is available in the queue (`queue_id`).
+
+
+## Timers
+
+### Functions exposed by the host environment
+
+#### `proxy_create_timer`
+
+* params:
+  - `i32 (uint32_t) period`
+  - `i32 (bool) one_time`
+  - `i32 (uint32_t*) return_timer_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Creates a new timer. When a timer is created as a one-time alarm (`one_time`),
+then callback will be called only once after `period` milliseconds. Otherwise,
+the callback is going to be called every `period` milliseconds until the timer
+is deleted using `proxy_delete_timer` with the returned unique timer identifier
+(`return_timer_id`).
+
+Returns `Ok` on success.
+
+Returns `InvalidMemoryAccess` when `return_timer_id` points to invalid memory
+address.
+
+#### `proxy_delete_timer`
+
+* params:
+  - `i32 (uint32_t) timer_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Deletes previously created timer (`timer_id`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the timer (`timer_id`) doesn't exist.
+
+### Callbacks optionally exposed by the Wasm module
+
+#### `proxy_on_timer_ready`
+
+* params:
+  - `i32 (uint32_t) timer_id`
+* returns:
+  - none
+
+Called after configured period by the timer (`timer_id`) created using
+`proxy_create_timer`.
+
+
+## Stats/metrics
```

Comment:
Where did we land on the differences between counters/gauges/histograms? Do we need a histogram API? It's not clear to be from reading the spec how a generalized API makes sense, or it seems like it should be more clear about what APIs are supported on what types and otherwise will return an error.